refactor(linter): improve `eslint/no-redeclare`

shulaoda opened the pull request:
Related to #8237 

- Improve the documentation
- ~~Align diagnostic format with `eslint`~~
- Fix errors exposed after adding test cases

camchenry reviewed (approved):
I kind of like showing the prior definition on the lint rule, but this is probably faster and aligned with ESLint I suppose

Sysix reviewed (commented):
I like the old diagnostic more too.

`builtinGlobals` should only look for following globals:
https://github.com/oxc-project/javascript-globals/blob/979277eeecf707f77ca7a5fd5ee55d3eef34fed5/src/lib.rs#L10-L72

shulaoda commented:
### Merge activity

* **Mar 22, 9:17 PM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

shulaoda committed:
refactor(linter): improve `eslint/no-redeclare` (#9976)

Related to #8237

- Improve the documentation
- ~~Align diagnostic format with `eslint`~~
- Fix errors exposed after adding test cases
